libkbfs/prefetcher.go

```diff
@@ -142,52 +393,83 @@ func (p *blockPrefetcher) calculatePriority(basePriority int,
+}
+
+func (p *blockPrefetcher) prefetchIndirectFileBlock(ctx context.Context,
+	parentBlockID kbfsblock.ID, b *FileBlock, kmd KeyMetadata,
+	lifetime BlockCacheLifetime) (numBlocks int, isTail bool) {
 	// Prefetch indirect block pointers.
-	p.log.CDebugf(context.TODO(), "Prefetching pointers for indirect file "+
-		"block. Num pointers to prefetch: %d", len(b.IPtrs))
 	startingPriority :=
 		p.calculatePriority(fileIndirectBlockPrefetchPriority, kmd.TlfID())
-	numBlocks := len(b.IPtrs)
-	doneCh := make(chan struct{}, numBlocks)
-	errCh := make(chan struct{}, numBlocks)
 	for i, ptr := range b.IPtrs {
-		_ = p.request(startingPriority-i, kmd, ptr.BlockPointer,
-			b.NewEmpty(), "", doneCh, errCh)
+		n, needNewFetch :=
+			p.recordPrefetchParent(ptr.BlockPointer.ID, parentBlockID)
+		numBlocks += n
+		if needNewFetch {
+			p.request(ctx, startingPriority-i, kmd,
+				ptr.BlockPointer, b.NewEmpty(), lifetime)
+		}
 	}
-	return doneCh, errCh, numBlocks
+	return numBlocks, len(b.IPtrs) == 0
 }
 
-func (p *blockPrefetcher) prefetchIndirectDirBlock(b *DirBlock,
-	kmd KeyMetadata) (<-chan struct{}, <-chan struct{}, int) {
+func (p *blockPrefetcher) prefetchIndirectDirBlock(ctx context.Context,
+	parentBlockID kbfsblock.ID, b *DirBlock, kmd KeyMetadata,
+	lifetime BlockCacheLifetime) (numBlocks int, isTail bool) {
 	// Prefetch indirect block pointers.
-	p.log.CDebugf(context.TODO(), "Prefetching pointers for indirect dir "+
-		"block. Num pointers to prefetch: %d", len(b.IPtrs))
 	startingPriority :=
 		p.calculatePriority(fileIndirectBlockPrefetchPriority, kmd.TlfID())
-	numBlocks := len(b.IPtrs)
-	doneCh := make(chan struct{}, numBlocks)
-	errCh := make(chan struct{}, numBlocks)
 	for i, ptr := range b.IPtrs {
-		_ = p.request(startingPriority-i, kmd, ptr.BlockPointer, b.NewEmpty(),
-			"", doneCh, errCh)
+		n, needNewFetch :=
+			p.recordPrefetchParent(ptr.BlockPointer.ID, parentBlockID)
+		numBlocks += n
+		if needNewFetch {
+			p.request(ctx, startingPriority-i, kmd,
+				ptr.BlockPointer, b.NewEmpty(), lifetime)
+		}
 	}
-	return doneCh, errCh, numBlocks
+	return numBlocks, len(b.IPtrs) == 0
 }
 
-func (p *blockPrefetcher) prefetchDirectDirBlock(ptr BlockPointer, b *DirBlock,
-	kmd KeyMetadata) (<-chan struct{}, <-chan struct{}, int) {
-	p.log.CDebugf(context.TODO(), "Prefetching entries for directory block "+
-		"ID %s. Num entries: %d", ptr.ID, len(b.Children))
+func (p *blockPrefetcher) prefetchDirectDirBlock(ctx context.Context,
+	parentBlockID kbfsblock.ID, b *DirBlock, kmd KeyMetadata,
+	lifetime BlockCacheLifetime) (numBlocks int, isTail bool) {
 	// Prefetch all DirEntry root blocks.
 	dirEntries := dirEntriesBySizeAsc{dirEntryMapToDirEntries(b.Children)}
 	sort.Sort(dirEntries)
 	startingPriority :=
 		p.calculatePriority(dirEntryPrefetchPriority, kmd.TlfID())
-	numBlocks := 0
-	doneCh := make(chan struct{}, len(dirEntries.dirEntries))
-	errCh := make(chan struct{}, len(dirEntries.dirEntries))
+	totalNumBlocks := 0
```

Comment:
Yeah, understood.  My suggestion was more about using "entries" here, instead of "blocks", because it doesn't really correspond to blocks well.  Maybe `totalChildEntries`?

Reply:
Sounds great, thanks.